### PR TITLE
Fixes the attachInterrupt() fails for pin D7 on Boron

### LINKED ISSUE #1932: attachInterrupt() fails for pin D7 on Boron
### Expected Behavior

attachInterrupt(D7, myFunc, RISING) == true;

### Observed Behavior

attachInterrupt(D7, myFunc, RISING) == false;

### Steps to Reproduce

Using firmware 1.4.0, attempt to attach an interrupt to D7 on a Boron device and it will return false.

I'm almost positive this is due to legacy (electron specific if my research is correct) code in spark_wiring_interrupts.cpp which has this catch in the attachInterrupt() definition:

```
#if Wiring_Cellular == 1
  /* safety check that prevents users from attaching an interrupt to D7
   * which is shared with BATT_INT_PC13 for power management */
  if (pin == D7) return false;
#endif 
```
And while the docs for the Electron clearly label this pin as off limits for that device, the docs for Boron state:

```
All A and D pins (including TX, RX, and SPI) on Gen 3 (mesh) devices can be used for interrupts...
```
I'm fairly certain Wiring_Cellular still gets set on the Boron though.

### PR DESCRIPTION
**Pull request by artyLee**
<details>
  <summary><i>submission notes</i></summary>

```
**Important:** Please sanitize/remove any confidential info like usernames, passwords, org names, product names/ids, access tokens, client ids/secrets, or anything else you don't wish to share.

Please Read and Sign the Contributor License Agreement ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md)).

You may also delete this submission notes header if you'd like. Thank you for contributing!
```
</details>

### Problem

The attachInterrupt() fails for pin D7 on Boron

### Solution

Fixes step:
1. Change HAL_Interrupts_Attach return type to int
2. Modify all HAL_Interrupts_Attach() implementations for all the platforms with that change and return 0 in case of success.
3. Modify hal_dynalib_gpio and also change the return type for HAL_Interrupts_Attach to int
4. Modify Gen 2 HAL_Interrupts_Attach() in hal/src/stm32f2xx/interrupts_hal.c to return SYSTEM_ERROR_NOT_ALLOWED (from system_error.h) on Electron if passed pin is D7
5. Modify wiring atttachInterrupt to check the return code from HAL_Interrupts_Attach and return true if it's 0 and false otherwise

### Steps to Test

Tested on the Boron and Electron. Use the following Example App to test. 

### Example App

```c
#include "application.h"
/*
Using a LED and 1k resistor connect the A0 pin and 3.3v, when the D7 interrupt happen the LED will toggle blink.
*/
SYSTEM_MODE(MANUAL);
void blink(void);
int ledPin = A0;
volatile int state = LOW;

void setup()
{
  Serial.begin(115200);
  pinMode(ledPin, OUTPUT);
  pinMode(D7, INPUT_PULLUP);
  attachInterrupt(D7, blink, CHANGE);
}

void loop()
{
  digitalWrite(ledPin, state);
}

void blink()
{
  state = !state;
  Serial.println("interrupt happened");
}


```

### References

- Closes #1932

---

### Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [x] Problem and Solution clearly stated
- [x] Run unit/integration/application tests on device
- [ ] Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

File: hal/inc/interrupts_hal.h [hunk 1]
```diff
@@ -91,9 +91,9 @@ extern "C" {
 
 void HAL_Interrupts_Init(void);
 void HAL_Interrupts_Uninit(void);
-void HAL_Interrupts_Attach(uint16_t pin, HAL_InterruptHandler handler, void* data, InterruptMode mode, HAL_InterruptExtraConfiguration* config);
-void HAL_Interrupts_Detach(uint16_t pin);
-void HAL_Interrupts_Detach_Ext(uint16_t pin, uint8_t keepHandler, void* reserved);
+int  HAL_Interrupts_Attach(uint16_t pin, HAL_InterruptHandler handler, void* data, InterruptMode mode, HAL_InterruptExtraConfiguration* config);
```
Comment (avtolstoy) [inline hal/inc/interrupts_hal.h]: Nitpicking: no need for two spaces :)

File: hal/src/nRF52840/interrupts_hal.cpp [hunk 2]
```diff
@@ -141,25 +142,26 @@ void HAL_Interrupts_Attach(uint16_t pin, HAL_InterruptHandler handler, void* dat
...
     if (keepHandler) {
         // This pin is used, don't detach it
-        return;
+        return SYSTEM_ERROR_INVALID_STATE;
     }
 
     Hal_Pin_Info* PIN_MAP = HAL_Pin_Map();
     if (PIN_MAP[pin].exti_channel == EXTI_CHANNEL_NONE) {
-        return;
+        return SYSTEM_ERROR_NOT_SUPPORTED;
```
Comment (avtolstoy) [inline hal/src/nRF52840/interrupts_hal.cpp]: This should be `SYSTEM_ERROR_INVALID_STATE`.

File: hal/src/nRF52840/interrupts_hal.cpp [hunk 3]
```diff
@@ -141,25 +142,26 @@ void HAL_Interrupts_Attach(uint16_t pin, HAL_InterruptHandler handler, void* dat
...
 
-void HAL_Interrupts_Detach(uint16_t pin) {
-    HAL_Interrupts_Detach_Ext(pin, 0, NULL);
+int HAL_Interrupts_Detach(uint16_t pin) {
+    return HAL_Interrupts_Detach_Ext(pin, 0, NULL);
 }
 
-void HAL_Interrupts_Detach_Ext(uint16_t pin, uint8_t keepHandler, void* reserved) {
+int HAL_Interrupts_Detach_Ext(uint16_t pin, uint8_t keepHandler, void* reserved) {
     if (keepHandler) {
```
Comment (avtolstoy) [inline hal/src/nRF52840/interrupts_hal.cpp]: This is actually a bug, and `keepHandler` isn't correctly handled. I'll commit to this branch to fix this.

File: hal/src/stm32f2xx/interrupts_hal.c [hunk 4]
```diff
@@ -80,8 +81,13 @@ static exti_state exti_saved_state = {0};
 
 /* Private function prototypes -----------------------------------------------*/
 
-void HAL_Interrupts_Attach(uint16_t pin, HAL_InterruptHandler handler, void* data, InterruptMode mode, HAL_InterruptExtraConfiguration* config)
+int HAL_Interrupts_Attach(uint16_t pin, HAL_InterruptHandler handler, void* data, InterruptMode mode, HAL_InterruptExtraConfiguration* config)
 {
+#if PLATFORM_ID == PLATFORM_ELECTRON_PRODUCTION
+  /* safety check that prevents users from attaching an interrupt to D7
+   * which is shared with BATT_INT_PC13 for power management */
+  if (pin == D7) return SYSTEM_ERROR_NOT_ALLOWED;
```
Comment (avtolstoy) [inline hal/src/stm32f2xx/interrupts_hal.c]: Nitpicking: formatting.
```suggestion
  if (pin == D7) {
      return SYSTEM_ERROR_NOT_ALLOWED;
  }
```

File: hal/src/stm32f2xx/interrupts_hal.c [hunk 5]
```diff
@@ -184,16 +190,22 @@ void HAL_Interrupts_Attach(uint16_t pin, HAL_InterruptHandler handler, void* dat
...
 }
 
 
-void HAL_Interrupts_Detach_Ext(uint16_t pin, uint8_t keepHandler, void* reserved)
+int HAL_Interrupts_Detach_Ext(uint16_t pin, uint8_t keepHandler, void* reserved)
 {
+#if PLATFORM_ID == PLATFORM_ELECTRON_PRODUCTION
+  /* safety check that prevents users from attaching an interrupt to D7
+   * which is shared with BATT_INT_PC13 for power management */
+  if (pin == D7) return SYSTEM_ERROR_NOT_ALLOWED;
```
Comment (avtolstoy) [inline hal/src/stm32f2xx/interrupts_hal.c]: Nitpicking: formatting.
```suggestion
  if (pin == D7) {
      return SYSTEM_ERROR_NOT_ALLOWED;
  }
```

File: wiring/src/spark_wiring_interrupts.cpp [hunk 6]
```diff
@@ -107,12 +102,9 @@ bool attachInterrupt(uint16_t pin, raw_interrupt_handler_t handler, InterruptMod
  *******************************************************************************/
 void detachInterrupt(uint16_t pin)
```
Comment (avtolstoy) [inline wiring/src/spark_wiring_interrupts.cpp]: Let's change this function signature to also return `true`/`false` to propagate the error.

**Comment by avtolstoy** [comment]
@artyLee Please also fix these functions for `newhal` platform, as the CI build for it failed.

File: wiring/src/spark_wiring_interrupts.cpp [hunk 7]
```diff
@@ -107,12 +102,9 @@ bool attachInterrupt(uint16_t pin, raw_interrupt_handler_t handler, InterruptMod
...
-#if Wiring_Cellular == 1
-    /* safety check that prevents users from detaching an interrupt from
-     * BATT_INT_PC13 for power management which is shared with D7 */
-    if (pin == D7) return;
-#endif
-    HAL_Interrupts_Detach(pin);
+    if (SYSTEM_ERROR_NONE != HAL_Interrupts_Detach(pin)) {
+        return;
+    }
     delete handlers[pin];
```
Comment (avtolstoy) [inline wiring/src/spark_wiring_interrupts.cpp]: Let's also add an `if` check here:
suggestion```
if (handlers[pin]) {
    delete handlers[pin];
    handlers[pin] = nullptr;
```

File: hal/inc/interrupts_hal.h (hunk 1 above)
Comment (artyLee) [reply hal/inc/interrupts_hal.h]: Ok, two spaces just for column alignment. I will follow your suggetion.

File: hal/src/nRF52840/interrupts_hal.cpp (hunk 2 above)
Comment (avtolstoy) [reply hal/src/nRF52840/interrupts_hal.cpp]: I fixed this one.

File: hal/src/nRF52840/interrupts_hal.cpp (hunk 3 above)
Comment (avtolstoy) [reply hal/src/nRF52840/interrupts_hal.cpp]: Fixed in https://github.com/particle-iot/device-os/pull/1939/commits/1e6ad02d278d05dfecfe86180a1da71f9cb5919e

File: hal/src/nRF52840/interrupts_hal.cpp (hunk 2 above)
Comment (artyLee) [reply hal/src/nRF52840/interrupts_hal.cpp]: Fixed

File: hal/src/nRF52840/interrupts_hal.cpp (hunk 3 above)
Comment (artyLee) [reply hal/src/nRF52840/interrupts_hal.cpp]: Great, thanks!

File: hal/src/stm32f2xx/interrupts_hal.c (hunk 4 above)
Comment (artyLee) [reply hal/src/stm32f2xx/interrupts_hal.c]: Fixed this formatting.

File: wiring/src/spark_wiring_interrupts.cpp (hunk 6 above)
Comment (artyLee) [reply wiring/src/spark_wiring_interrupts.cpp]: Fixed this function signature return type `bool`.

File: hal/src/stm32f2xx/interrupts_hal.c (hunk 5 above)
Comment (artyLee) [reply hal/src/stm32f2xx/interrupts_hal.c]: Fixed this formatting.

File: wiring/src/spark_wiring_interrupts.cpp (hunk 7 above)
Comment (artyLee) [reply wiring/src/spark_wiring_interrupts.cpp]: Fixed this line code for using`if` check

**Comment by avtolstoy** [comment]
@artyLee Would you also mind adding a test to `user/tests/wiring/no_fixture/interrupts.cpp` for Electron specifically, to test that `attachInterrupt`/`detachInterrupt` will return `false` with `D7`?

**Comment by artyLee** [comment]
> @artyLee Would you also mind adding a test to `user/tests/wiring/no_fixture/interrupts.cpp` for Electron specifically, to test that `attachInterrupt`/`detachInterrupt` will return `false` with `D7`?

OK, I have added a test code, if you have time please review. Thanks!

**Pull request merged**